### ISSUE TITLE
[DOOR Bugs] `TypeHint(Union[...])` comparators busted in edge cases

### ISSUE THREAD

**Issue by wesselb** [beartype/beartype#198]
Hey @leycec!

I've started playing with the goodies of @beartype, and `TypeHint` has been working fantastically. :)

However, I've been wondering about a particular behaviour involving `Union`s.

Check this out:

```python
from typing import Union
from beartype.door import TypeHint
from numbers import Number


class A:
    pass

class B(A):
    pass
```

```python
>>> TypeHint(Union[A]) >= TypeHint(Union[A, B])
False
```

Should this be `True`? In fact, I'm inclined to believe that `TypeHint(Union[A]) == TypeHint(Union[A, B])` in this case.

Also,

```python
>>> TypeHint(Union[Number, int]) >= TypeHint(Union[Number, float])
True

>>> TypeHint(Union[Number, int]) <= TypeHint(Union[Number, float])
True

>>> TypeHint(Union[Number, int]) == TypeHint(Union[Number, float])
False
```

What are your thoughts on the above behaviour?

**Comment by leycec** [comment]
**Blargh!** This some *sus af tbh* behaviour there. On the one hand, one of these comparisons seems "possibly okay" for competing definitions of "possibly" and "okay":

```python
>>> TypeHint(Union[Number, int]) == TypeHint(Union[Number, float])
False
```

Since `int` and `float` are incommensurable (i.e., have no subclass relationship ignoring [PEP 484's implicit numeric tower](https://peps.python.org/pep-0484/#the-numeric-tower)), we would expect all comparisons between `int` and `false` to evaluate to `False`. So, that *sorta* makes sense. These two, however, do not:

```python
>>> TypeHint(Union[Number, int]) >= TypeHint(Union[Number, float])
True
>>> TypeHint(Union[Number, int]) <= TypeHint(Union[Number, float])
True
```

:vomiting_face: 

For the same reason, these two comparisons should also evaluate to `False`. They do not. That's bad. In response, the black hole at the heart of the `beartype.door` API voraciously begins devouring everything in its stellar neighbourhood.

And *this* is undeniably awful:

```python
>>> TypeHint(Union[A]) >= TypeHint(Union[A, B])
False
```

:vomiting_face: :vomiting_face: :vomiting_face: 

That's the exact opposite of what you'd expect! Intuitively, `Union[A]` *seems* like it should definitely be a subhint of `Union[A, B]`. The former is commensurable with the latter but matches substantially fewer (in the "countably infinitely fewer" sense of "fewer") type hints than the latter. If these words I am typing are true, we would expect that comparison to instead evaluate to `True`: e.g.,

```python
# Listen and obey, @beartype! *ROAR* Bad bear, bad bear! *ROAR, ROAR*
>>> TypeHint(Union[A]) >= TypeHint(Union[A, B])
True
```

# A Heroic Plan for Fixing that Which Has Been Broken

Sadly, `beartype 0.12.0` will probably preserve this broken behaviour. Scarce open-source time, thy name is Christmas. If you'd like to delve into this a bit, though, that'd be a wondrous Yuletide miracle! The relevant code resides in:

* The public `beartype.door._doorcls.TypeHint.is_subhint()` method, which is where the illicit and poorly maintained fireworks begin exploding. This method internally defers to...
* The private `beartype.door._doorcls.TypeHint._is_le_branch()` method, where most of the heavy-lifting in type hint comparisons resides.
* The private `beartype.door._proposal.doorpep484604.

Specifically, this code snippet strikes me as extremely suspicious:

```python
class TypeHint(...):
    def is_subhint(...):
        ...

        # Return true only if this type hint is a subhint of *ANY* branch of the
        # passed type hint.
        return any(
            self._is_le_branch(other_branch)
            for other_branch in other._branches
        )
```

**Yikes!** What are you even doing there, @beartype? :face_exhaling:

**Comment by leycec** [comment]
**Curioser** and **curioser.** I didn't entirely understand the prior logic for `UnionTypeHint.is_subhint()`, so I refactored everything according to Dr. Demento's Inscrutable Law of "Less Code Must Be Better." But it turns out that I didn't really need to do that at all, as it changed nothing.

After having cogitated deeply about the true meaning of unions at Christmas while creeping about the cottage at midnight, I have confirmed to at least myself that:

1. This is goodness incarnate:

```python
>>> TypeHint(Union[Number, int]) >= TypeHint(Union[Number, float])
True
>>> TypeHint(Union[Number, int]) <= TypeHint(Union[Number, float])
True
```

Since `Number` is the superclass of both `int` and `float`, `Union[Number, int]` and `Union[Number, float]` semantically reduce to just `Number`. Since `TypeHint(Number) >= TypeHint(Number)` and `TypeHint(Number) <= TypeHint(Number)` are both trivially true, the above results hold. Thus do we thank you, Santa Bearclaws.

2. This is badness personified, however:

```python
>>> TypeHint(Union[Number, int]) == TypeHint(Union[Number, float])
False
```

By the same reasoning, that should evaluate to `True` as well. It currently doesn't. Why? Because our `UnionTypeHint` subclass isn't redefining the `__eq__()` dunder method. It should be. Rudolf the Bug-sniffing Reindeer: *you know what to do here.*

3. This is badness in the flesh, too:

```python
>>> TypeHint(Union[A]) >= TypeHint(Union[A, B])
False
```

Exactly as you suggest, @wesselb, both that *and* the equality `TypeHint(Union[A]) == TypeHint(Union[A, B])` should absolutely evaluate to `True`. You're right about everything, as always!

Thankfully, I can explain exactly what's going wrong here, too. It may not look like it, but `Union[A] is A`. Literally. This is not semantics. Python's standard library tries to be clever by reducing a union containing only one type hint to just that type hint at runtime. For clarity, let's rewrite the above comparison using that explicit reduction as well as the `<=` operator:

```python
>>> TypeHint(Union[A, B]) <= TypeHint(A)
False
```

**Ah-ha!** Now we see exactly what's gone wrong here. As currently implemented, the `UnionTypeHint.is_subhint()` method (underlying the `<=` operator) only supports comparisons between *union* type hints. Clearly, that's insufficient. No one ever expected an edge case union type hint like this. A union absolutely *can* be a subhint of a non-union, as the above example capably demonstrates.

So, we need to also generalize the `UnionTypeHint.is_subhint()` method to transparently support comparisons against non-union type hints as well. How? By just treating a non-union type hint as a union of only that type hint, which is really just reversing the reduction that Python itself uses. *Viola!*

Santa Bearclaws, you may now eat all the gluten-free cookies.

**Comment by wesselb** [comment]
As usual, yet another fantastic reply by @leycec. :) That `Union[A] is A` has tricked me before too!

I'm glad you've gotten to the bottom of this so super quickly. Thanks again for your hard work!! We all love @beartype 😃

(A @beartype-powered Plum 2.0 should now be coming up soooooon!)

**Comment by leycec** [comment]
**Aww.** You're so welcome. Big wintry hugs from the ~~enslaved~~ happy QA elves at @beartype to you and your family this Christmas. Plum is equally amazing and about to blow even more of our socks off! :socks: :christmas_tree: :santa: 

Which is a good thing on multiple counts. My wife insists my socks simply *must* be cleaned. Much to everyone's detriment, I keep resisting this sound advice. *"Viva la poor hygiene."*

**Comment by wesselb** [comment]
Big beary hugs and a Merry Christmas to you and your family too! 🎅

Let's make sure those socks get blown off soon then haha. :P 🧦

**Comment by leycec** [comment]
**Socks officially blown!** Mine, that is. Resolving these edge cases was *super*-fun but *super*-non-trivial. It's like we're building out a theorem prover built on the unstable back of the `typing` module capable of solving Fermat's Last QA Problem. The DOOR API is rapidly transforming into this monolithic type hint calculus monstrosity capable of deciding that 42 might, after all, be the meaning of Python.

In short, this has now been resolved by c449dfc – to be released with our upcoming `beartype 0.12.0` mic drop just as soon as I catch up on six month's worst of loss sleep. Wish me luck, boys! I'm crashin' hard tonight. :sleeping:

**Comment by wesselb** [comment]
Thank you, @leycec!! This is super great. :)
